lib/index.js

```diff
@@ -153,36 +153,72 @@ export function configureUnifiedLanguageServer(
    * @returns {Promise<VFile[]>}
    */
   function processDocuments(textDocuments, alwaysStringify = false) {
-    return new Promise((resolve, reject) => {
-      engine(
-        {
-          alwaysStringify,
-          files: textDocuments.map((document) => lspDocumentToVfile(document)),
-          ignoreName,
-          packageField,
-          pluginPrefix,
-          plugins,
-          processor,
-          quiet: false,
-          rcName,
-          silentlyIgnore: true,
-          streamError: new PassThrough(),
-          streamOut: new PassThrough()
-        },
-        (error, _, context) => {
-          // An error never occur and can’t be reproduced. Thus us ab internal
-          // error in unified-engine. If a plugin throws, it’s reported as a
-          // vfile message.
-          /* c8 ignore start */
-          if (error) {
-            reject(error)
-          } else {
-            resolve((context && context.files) || [])
-          }
-          /* c8 ignore stop */
-        }
-      )
-    })
+    // LSP uses `file:` URLs (hrefs), `unified-engine` expects a paths.
+    // `process.cwd()` does not add a final slash, but `file:` URLs often do.
+    const workspacesAsPaths = workspaces
+      ? workspaces.map((d) => fileURLToPath(d.replace(/\/$/, '')))
+      : [process.cwd()]
+    /** @type {Record<string, Array<VFile>>} */
+    const workspacePathToFiles = {}
+
+    for (const textDocument of textDocuments) {
+      const file = lspDocumentToVfile(textDocument)
+      const [cwd] = workspacesAsPaths
+        // Every workspace that includes the document.
+        .filter((d) => file.path.slice(0, d.length + 1) === d + path.sep)
+        // Sort the longest (closest to the file) first.
+        .sort((a, b) => b.length - a.length)
+
+      // This presumably should not occur: a file outside a workspace.
+      // So ignore the file.
+      /* c8 ignore next */
+      if (!cwd) continue
+
+      if (!(cwd in workspacePathToFiles)) {
+        workspacePathToFiles[cwd] = []
+      }
+
+      workspacePathToFiles[cwd].push(file)
+    }
+
+    return Promise.all(
+      Object.keys(workspacePathToFiles).map((cwd) => {
```

Comment:
Since both keys and values are used, `Object.entries` can be used to avoid the need to look up the value.

Since there’s another suggestion to use a map, the following construct can be used:

```js
return Promise.all(Array.from(workspacePathToFiles.entries(), ([cwd, files]) => 
  new Promise((resolve, reject) => {
    engine(
      {
        alwaysStringify,
        cwd,
        files,
        ignoreName,
        packageField,
        pluginPrefix,
        plugins,
        processor,
        quiet: false,
        rcName,
        silentlyIgnore: true,
        streamError: new PassThrough(),
        streamOut: new PassThrough()
      },
      (error, _, context) => {
        // An error never occurs and can’t be reproduced. Thus it’s an internal
        // error in unified-engine. If a plugin throws, it’s reported as a
        // vfile message.
        /* c8 ignore start */
        if (error) {
          reject(error)
        } else {
          resolve((context && context.files) || [])
        }
        /* c8 ignore stop */
      }
    )
  })
))
```